File: Board.py
class Tile:
    num = None
    candidates = []
    row = 1
    column = 5
    square = 3
    def __init__(self, val, board=None):
        self.board = board
        if val not in range(1,10):
            self.num = None
            self.candidates = list(range(1,10))
        else:
            self.num = val
    def __repr__(self):
        if self.board:
            return f"{self.board.name}: {str(self.num)}"
        else:
            return str(self.num)
    def update(self):
        if len(self.candidates) == 1:
            self.num = self.candidates[0]
            self.candidates = []
            self.row.in_row.append(self.num)
            self.column.in_col.append(self.num)
            self.square.in_square.append(self.num)
class Board:
    tiles = None
    rows =[]
    cols = []
    name = 'Aniruddh\'s Sudoku Board'
    def __init__(self, tiles):
        self.squares = [Square() for _ in range(9)]
        for i in range(0,len(tiles)):
            for j in range(0,len(tiles[i])):
                tiles[i][j] = Tile(tiles[i][j], self)
                if i < 3:
                    if j < 3:
                        self.squares[0].add(tiles[i][j])
                    elif j < 6:
                        self.squares[1].add(tiles[i][j])
                    else:
                        self.squares[2].add(tiles[i][j])
                elif i < 6:
                    if j < 3:
                        self.squares[3].add(tiles[i][j])
                    elif j < 6:
                        self.squares[4].add(tiles[i][j])
                    else:
                        self.squares[5].add(tiles[i][j])
                else:
                    if j < 3:
                        self.squares[6].add(tiles[i][j])
                    elif j < 6:
                        self.squares[7].add(tiles[i][j])
                    else:
                        self.squares[8].add(tiles[i][j])
        self.tiles = tiles
        for i in range(0, len(self.tiles)):
            r = Row(tiles[i])
            self.rows.append(r)
        self.cols = [Column([my_tiles[i][j] for i in range(len(my_tiles))])for j in range(len(my_tiles))]
    def __repr__(self):
        out = 'Aniruddh\'s Sudoku Board \n'
        for i in range(0, len(self.tiles)):
            for j in range(0, len(self.tiles[i])):
                if self.tiles[i][j].num:
                    out += str(self.tiles[i][j].num)
                    out += ' '
                else:
                    out += '_ '
            out += '\n'
        return out
    def check_complete(self):
        for r in self.tiles:
            for t in r:
                if t.num == None:
                    return False
        return True
class Row:
    tiles = []
    def __init__(self, tiles):
        self.in_row = []
        for tile in tiles:
            tile.row = self
            if tile.num:
                self.in_row.append(tile.num)
        self.tiles = tiles
    def __repr__(self):
        out = ''
        for tile in self.tiles:
            if tile.num:
                out += str(tile.num)
                out += ' '
            else:
                out += '_ '
        return out
class Column:
    tiles = []
    def __init__(self, tiles):
        self.in_col = []
        for tile in tiles:
            tile.column = self
            if tile.num:
                self.in_col.append(tile.num)
        self.tiles = tiles
    def __repr__(self):
        out = ''
        for tile in self.tiles:
            if tile.num:
                out += str(tile.num)
                out += '\n'
            else:
                out += '_\n'
        return out
class Square:
    def __init__(self):
        self.in_square = []
        self.tiles = []
    def add(self, tile):
        self.tiles.append(tile)
        tile.square = self
        if tile.num:
            self.in_square.append(tile.num)
    def __repr__(self):
        out = ''
        for i in range(len(self.tiles)):
            if self.tiles[i].num:
                out += str(self.tiles[i].num)
            else:
                out += '_'
            if (i+1) % 3 == 0:
                out += '\n'
            else:
                out += ' '
        return out

# 31 none
my_tiles = [
[1,0,0,0,0,0,6,0,5],
[0,0,2,0,1,7,0,0,9],
[4,0,0,2,0,0,7,0,8],
[7,0,0,4,0,0,1,0,3],
[8,0,4,0,0,0,0,0,0],
[0,2,5,0,0,9,0,0,0],
[0,6,0,0,3,0,2,0,0],
[0,0,1,5,0,0,3,0,0],
[0,0,0,6,9,8,0,7,0]]
my_board = Board(my_tiles)

File: Board.java
//public class Board {
//    Tile[][] _tiles;
//    Row[] _rows;
//    Column[] _cols;
//    Square[] _squares;
//    String name = "Aniruddh's Sudoku Board";
//    boolean updated = true;
//    boolean incomplete = true;
//
//    public Board(Tile[][] tiles, int dim) {
//        _rows = new Row[dim];
//        _cols = new Column[dim];
//        _squares = new Square[dim];
//        for(int i = 0; i < tiles.length; i += 1) {
//            for(int j = 0; j < tiles[i].length; j += 1) {
//                tiles[i][j] = new Tile(tiles[i][j]);
////                need to add to a square
//            }
//        }
//        _tiles = tiles;
//        for (int i = 0; i < _tiles.length; i += 1) {
//            Row r = new Row(_tiles[i]);
//            _rows[i] = r;
//        }
//    }
//}

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;

public class Board {
    class Row extends Group {
        Row (Tile[] tiles, int dim) {
            super(tiles, dim, 'r');
        }
    }

    class Column extends Group {
        Column (Tile[] tiles, int dim) {
            super(tiles, dim, 'c');
        }
    }

    class Square extends Group {
        Square (int dim) {
            super(dim, 's');
        }
    }
    Tile[][] _tiles = null;
    Row[] _rows;
    Column[] _cols;
    Square[] _squares;
    String name = "Aniruddh's Sudoku Board";
    boolean updated = true;
    boolean incomplete = true;
    public Board(int[][] tiles, int dim) {
        _squares = new Square[dim];
        _rows = new Row[dim];
        _cols = new Column[dim];
        for (int i = 0; i < dim; i += 1) {
            _squares[i] = new Square(dim);
        }
        _tiles = new Tile[tiles.length][];
        for (int i = 0; i < tiles.length; i += 1) {
            _tiles[i] = new Tile[tiles[i].length];
            for (int j = 0; j < tiles[i].length; j += 1) {
                _tiles[i][j] = new Tile(tiles[i][j], this);
                int sqSize = (int) Math.sqrt(dim);
                int sq = (sqSize * (i / sqSize)) + (j / sqSize);
                _squares[sq].addTile(_tiles[i][j]);
                _tiles[i][j]._square = _squares[sq];
            }
            Row myRow = new Row(_tiles[i], dim);
            _rows[i] = myRow;
        }
        for(int j = 0; j < tiles[0].length; j += 1) {
            Tile[] myCol = new Tile[dim];
            for(int i = 0; i < tiles.length; i += 1) {
                myCol[i] = _tiles[i][j];
            }
            _cols[j] = new Column(myCol, dim);
        }
    }
    public boolean isComplete() {
        for (Tile[] row : _tiles) {
            for (Tile t : row) {
                if (t._num == 0) {
                    incomplete = true;
                    return false;
                }
            }
        }
        incomplete = false;
        return true;
    }

    public void checkGroup(Group g) {
        for (Tile t : g.get_tiles()) {
            if (t.candidates.size() > 0 && t._num == 0) {
                t.candidates.removeIf(el -> (g.inGroup.containsKey(el)));
                int dummy = 0;
            }
        }
    }
    public void updateCandidates() {
        for (Row r : _rows) {
            checkGroup(r);
        }
        for (Column c : _cols) {
            checkGroup(c);
        }
        for (Square s : _squares) {
            checkGroup(s);
        }
        int dummy = 0;

    }
    public ArrayList<Integer> uniqueInGroup(Group g) {
        ArrayList<HashSet<Integer>> nested = new ArrayList();
        for (Tile t : g.get_tiles()) {
            if (t._num == 0) {
                nested.add(t.candidates);
            }
        }
        ArrayList<Integer> allCandidates = new ArrayList<>();
        for (HashSet<Integer> nest : nested) {
            allCandidates.addAll(nest);
        }
        HashMap<Integer, Integer> candidateCounts = convertToMap(allCandidates);
        ArrayList<Integer> uniqueCandidates = new ArrayList<>();
        for (Integer candidate : candidateCounts.keySet()) {
            if (candidateCounts.get(candidate) == 1) {
                uniqueCandidates.add(candidate);
            }
        }
        return uniqueCandidates;
    }
    public HashMap<Integer, Integer> convertToMap(ArrayList<Integer> elements) {
        HashMap<Integer, Integer> map = new HashMap<>();
        for (Integer element : elements) {
            if(map.containsKey(element)) {
                map.put(element, map.get(element) + 1);
            } else {
                map.put(element, 1);
            }
        }
        return map;
    }

    public void updateTiles() {
        for (Tile[] row : _tiles) {
            for (Tile tile : row) {
                tile.update();
            }
        }
    }
    public Board updateBoard() {
        updated = false;
        for (Tile[] row : _tiles) {
            for (Tile tile : row) {
                tile.resetCandidates();
            }
        }
        updateCandidates();
        updateTiles();
        dump();
        isComplete();
        return this;
    }
    public void dump() {
        String out = "Aniruddh's Sudoku Board \n";
        for (int i = 0; i < _tiles.length; i ++) {
            for (int j = 0; j < _tiles[i].length; j++) {
                if (_tiles[i][j]._num != 0) {
                    out += Integer.toString(_tiles[i][j]._num);
                    out += " ";
                } else {
                    out += "_ ";
                }
            }
            out += "\n";
        }
        System.out.println(out);
    }
    public void solve() {
        int iters = 0;
//        "The smart approach"
        while (this.incomplete &&  this.updated && iters< 25) {
            updateBoard();
            iters += 1;
        }
//        Backtracking - brute force

    }

    public void backtrack() {
        for (Tile[] row : _tiles) {
            for (Tile t : row) {
                if (t._num == 0 && t.candidates != null && t.candidates.size() > 0) {
                    for (int candidate : t.candidates) {
                        t._num = candidate;
                        t.addToGroup();
                    }
                }
            }
        }
    }
}